cmd/kubeadm/app/phases/selfhosting/podspec_mutation.go

```diff
@@ -96,20 +96,25 @@ func addNodeSelectorToPodSpec(podSpec *v1.PodSpec) {
 // setMasterTolerationOnPodSpec makes the Pod tolerate the master taint
 func setMasterTolerationOnPodSpec(podSpec *v1.PodSpec) {
 	if podSpec.Tolerations == nil {
-		podSpec.Tolerations = []v1.Toleration{kubeadmconstants.MasterToleration}
+		podSpec.Tolerations = []v1.Toleration{
+			kubeadmconstants.MasterToleration,
+			kubeadmconstants.CloudProviderUninitializedToleration,
+		}
 		return
 	}
 
-	podSpec.Tolerations = append(podSpec.Tolerations, kubeadmconstants.MasterToleration)
+	podSpec.Tolerations = append(podSpec.Tolerations,
+		kubeadmconstants.MasterToleration,
+		kubeadmconstants.CloudProviderUninitializedToleration)
 }
 
-// setHostIPOnPodSpec sets the environment variable HOST_IP using downward API
-func setHostIPOnPodSpec(podSpec *v1.PodSpec) {
+// setPodIPOnPodSpec sets the environment variable POD_IP using downward API
```

Comment:
Doesn't this have to be HOST_IP for self-hosted api-servers?  I don't see how it could possibly be POD_IP if you put a LB in front.

Reply:
The idea was that POD_IP is effectively the same as HOST_IP since we run this pod on host network. But @aoxn pointed out that the kubelet actually ends up using the same code path regardless to fetch the host IP which leads us back to the original problem where we can't use self hosted with external clouds because we rely on the cloud controllers to fill in the necessary information to fetch the node's host IP.

Reply:
I think there needs to be a discussion to address these sorts of issues in general, I'll bring it this up in cloud provider WG next week.

Reply:
@andrewsykim - I'll happily add you to the sig-cluster-lifecycle agenda tomorrow if you have a chance to come.

Reply:
Sure thing, I can drop by tomorrow!

Reply:
Did we make progress on this?

Reply:
The conclusion was that self hosting + kubeadm  with external clouds doesn't work as expected at the moment largely due to baked in assumptions from the kubelet (other components too but mostly kubelet). Steps forward are to document clearly that self hosted kubeadm with external clouds will likely not work. Cloud provider WG will be discussing potential solutions to this problem but it will take some time since we need to break down the kubelet a bit to make this work.